zcash_proofs/src/sapling/prover.rs

```diff
@@ -261,6 +262,11 @@ impl SaplingProvingContext {
         // Grab the `bvk` using DerivePublic.
         let bvk = PublicKey::from_private(&bsk, FixedGenerators::ValueCommitmentRandomness, params);
 
+        // XXX is this actually necessary or is it a sanity check? if the
+        // binding signatures are properly encapsulated, the raw components of
+        // the signature are not available, so this check is much more difficult
+        // to do.
+
```

Comment:
I'm not sure whether this check is actually necessary (because it's possible to construct a builder that has some invalid data or something), or whether it's just a sanity check and could be safely eliminated.

If the binding signatures are properly encapsulated, the raw components of the signature aren't available, so it's much more difficult to do.  But if it could be eliminated, there's also no need to propagate the `value_balance` parameter into this function.  And because there's no need to pass `params`, the only parameters would then be `sighash` and `self`.  But at that point the only purpose of the function would be to construct the binding key and use it for a signature, so `sighash` could also be eliminated by making the function just compute the binding key and return it for the caller to use in signing.

Reply:
Per @daira, this is a sanity check: 

![image](https://user-images.githubusercontent.com/44879/72643573-d5b0e500-3923-11ea-8877-15e766e167bf.png)

Reply:
Perhaps, then, the best way to refactor this would be to propagate `value_balance` as current but change the function to one that just returns a `Result<SecretKey<Binding>,_>`, with an `Ok` if the sanity check passes and an `Err` if it doesn't?